### DISCUSSION TITLE
Slot Polymorphism cloning technique

### DISCUSSION THREAD

**Discussion by andresin87** [adevinta/spark#577] [Architecture Decision Record]
# Slot Polymorphism cloning technique

## Status

![status](https://img.shields.io/badge/status-agree-green?style=plastic)

### SLOT

React allows you to pass children to a component by nesting them inside its JSX tag.

***Slot*** component inject its given props to that children applying a set of **[rules](https://sparkui.vercel.app/?path=/docs/components-slot--docs)**

### POLYMORPHISM
For applying React polymorphism there are 2 different ways:
- `as: (reactElement)`  prop: for element tag-names (strings), or ReactComponents
- `asChild: boolean` prop: in case of setting it to true, user can declare what element needs to look like the element with that flag and pass it as a children.

- Given:
	- Parent
```jsx
forwardRef({asChild, ...props}, forwardedRef) => {
	const Component = asChild ? Slot : 'div'
	return <Component {...props} ref={forwardedRef} />
}
```

	- Children
```jsx
forwardRef({asChild, ...props}, forwardedRef) => (
  <a
    {...props}
    ref={forwardedRef}
    data-prop="children"
  />
)
```

#### element `as` prop

- Given:
	- Parent
```jsx
forwardRef({as: Component = "div" , ...props}, forwardedRef) => (
  <Component
    {...props}
    ref={forwardedRef}
  />
}
```

	- Children
```jsx
forwardRef(props, forwardedRef) => (
  <a
    ref={forwardedRef}
    data-collision-prop="children"
    {...props}
    {/*** data-collision-prop="children" ***/}
  />
)
// The data collision prop can be added before of after the prop spreading and will have a different merging technique
```

When coding...

```jsx
<Parent
  as={Children}
  data-collision-prop="parent"
  data-other-prop="parent"
/>
```
...will generate as an HTML result: 

```HTML
<a  data-collision-prop="???" data-other-prop="parent" />
```

Is not possible to determine which is the result in case of prop declared collision without reading the implementation of the Children's prop assignation.

#### `asChild` prop boolean Flag

- Given:
	- Parent
```jsx
forwardRef({asChild , ...props}, forwardedRef) => {
  const Component = asChild ? Slot : 'div'
  return (
    <Component
	  ref={forwardedRef}
      {...props}
    />
  }
```

	- Children
```jsx
forwardRef(props, forwardedRef) => (
  <a
    ref={forwardedRef}
    data-collision-prop="children"
    {...props}
    {/*** data-collision-prop="children" ***/}
  />
)
// The data collision prop can be added before of after the prop spreading and will have THE SAME merging technique (slot)
```

When coding...

```jsx
<Parent
  asChild
  data-collision-prop="parent"
  data-other-prop="parent"
>
  <Children />
</Parent>
```
...will generate as an HTML result: 

```HTML
<a  data-collision-prop="children" data-other-prop="parent">
```
This functionality gives a deterministic centralized behavior (inside Slot logic)

### THE PROBLEM

Combining both techniques had a list of  problem when handling the reference `ref` prop

The reference (`ref` prop) in Slot component 

## Decision

What is the change that we're proposing and/or doing?

## Consequences

What becomes easier or more difficult to do because of this change?

**Comment by Powerplex** [comment]
Love the specs for Slot :) 

Regarding the problem at the end, I'm not sure if it is resolved with `asChild` or if you want to discuss it ?

**Reply by andresin87** [reply]
WIP

**Comment by andresz1** [comment]
Can this discussion be closed since we are already using `Slot` and `Slottable` @andresin87 ?

**Comment by Powerplex** [comment]
Feature is complete.
